### DISCUSSION TITLE
Emulating child-relations

### DISCUSSION THREAD

**Discussion by oliverw** [Azure/azure-mobile-apps#600] [Q&A] [answered]
One of the features I need to implement in our current project is one entity, lets call it `Task` to have a number of `Sub-Tasks`. The number of `Sub-Tasks` is curated by a human so the number is neither fixed nor incredibly high. Let's assume < 100.  There is also just one hierarchy level and no tree-like structure.

I've read the various tickets and discussions and concluded that implementing this using DB foreign keys on top of the SDK's own persistence layer is a bad idea (correct me if I'm wrong). The notion of inter-table relations just doesn't seem to fit with the current architecture of the Azure Mobile Apps - which is fine. 

One possible - albeit hacky smelling - approach would be to implementing this relation entirely client-side by maintaining a JSON document on a table entity property. Thoughts?

**Comment by adrianhall** [comment, marked as answer]
The hacky version smells too.

What you need to do is separate the tables on the client and implement the foreign key on the service side.  Let's say you have TodoList and TodoItem:

```csharp
public class TodoList : DatasyncClientData
{
  public string Name { get; set; }
  public string? Description { get; set; }
}

public class TodoItem : DatasyncClientData
{
  public string ListId { get; set; }
  public string Name { get; set; }
  public string Description { get; set; }
  public TodoItemState State { get; set; }
}
```

You can now link them manually on the client-side (by doing `TodoItems.Where(x => x.ListId == listId)...`) and you are responsible for:

* Ensuring you only have one thread when pushing the requests.
* Ensuring the TodoList creation happens before the associated TodoItem requests.

That should be enough to get your data models working.  You may have to use AutoMapper to map to database models onto DTOs that your datasync controllers use.

**Reply by oliverw** [reply]
Interesting. Thanks Adrian 👍

**Reply by oliverw** [reply]
One a side note: If you had self-referencing entities. Could it happen that the parent is pushed after the child entity on `PushItemsAsync`, leading to an exception on the backend?

**Reply by adrianhall** [reply]
Certainly possible.   We take pains to ensure that the ordering of the operations queue is appropriate, but throw that out if using threaded push or if you add collapsible entities into the queue.
